File: python/euler/bin/p0031.py
# project euler: problem 31


def compute(coins: list[int], target: int) -> str:
    tbl = [0] * (target + 1)
    tbl[0] = 1

    for c in coins:
        for i in range(c, target + 1):
            tbl[i] += tbl[i - c]

    return str(tbl[target])


def solve() -> str:
    return compute([1, 2, 5, 10, 20, 50, 100, 200], 200)

File: typescript/euler/bin/p0031.ts
// project euler: problem 31

import { range } from "../lib/util.ts";

export const compute = (coins: number[], target: number): string => {
  const tbl = new Array(target + 1);
  tbl.fill(0);
  tbl[0] = 1;

  for (const c of coins) {
    for (const i of range(c, target + 1)) {
      tbl[i] += tbl[i - c];
    }
  }

  return String(tbl[target]);
};

export const solve = (): string =>
  compute([1, 2, 5, 10, 20, 50, 100, 200], 200);
